[libc++] P3247R2: Deprecate `is_trivial(_v)`

frederick-vs-ja opened the pull request:
Requirements on character-like types are updated unconditionally, because `basic_string` does requires the default-constructibility. It might be possible to make `basic_string_view` support classes with non-public trivial default constructor, but this doesn't seem sensible.

libcxxabi's `ItaniumDemangle.h` is also updated to avoid deprecated features.

Fixes #118387.

philnik777 commented on libcxx/include/__iterator/aliasing_iterator.h:
```diff
@@ -45,7 +46,8 @@ struct __aliasing_iterator_wrapper {
 
-    static_assert(is_trivial<value_type>::value);
+    static_assert(is_trivially_default_constructible<value_type>::value);
+    static_assert(is_trivially_constructible<value_type>::value);
```
This should probably be `is_trivially_copyable`.

philnik777 commented on libcxx/test/libcxx/utilities/memory/util.smartptr/util.smartptr.shared/libcxx.control_block_layout.pass.cpp:
```diff
@@ -123,7 +123,9 @@ void test() {
   // Make sure both types have the same triviality (that has ABI impact since
   // it determined how objects are passed). Both should be non-trivial.
-  static_assert(std::is_trivial<New>::value == std::is_trivial<Old>::value, "");
+  static_assert((std::is_trivially_copyable<New>::value && std::is_trivially_default_constructible<New>::value) ==
+                    (std::is_trivially_copyable<Old>::value && std::is_trivially_default_constructible<Old>::value),
+                "");
```
We can probably enforce that `is_trivially_copyable<New> == is_trivially_copable<Old> && ...`.

philnik777 commented on libcxx/test/std/depr/depr.c.headers/stddef_h.compile.pass.cpp:
```diff
@@ -42,11 +42,13 @@ static_assert(std::is_integral<ptrdiff_t>::value, "");
+static_assert(std::is_trivially_default_constructible<max_align_t>::value, "");
 static_assert(std::is_standard_layout<max_align_t>::value, "");
-#  else
+#  if TEST_STD_VER < 26 // P3247R2
+static_assert(std::is_trivial<max_align_t>::value, "");
+#  endif
```
I don't think this is necessary. We're testing `max_align_t`, not the type traits.

frederick-vs-ja replied:
Yeah. But some existing uses of `is_pod` look weird. I think it's better to remove them, otherwise some test files will become weirder.

philnik777 commented on libcxx/test/std/language.support/support.types/byte.pass.cpp:
```diff
@@ -10,26 +10,34 @@
 // std::byte is not an integer type, nor a character type.
-// It is a distinct type for accessing the bits that ultimately make up object storage.
+// It is a distinct scoped enumeration type for accessing the bits that ultimately make up object storage.
 
```
Let's do the refactoring in a separate patch. Otherwise it's really not clear which parts are just refactoring and which parts are actually related to the paper.

philnik777 commented on libcxx/test/std/language.support/support.types/byte.pass.cpp:
```diff
@@ -10,26 +10,34 @@
-int main(int, char**) {
-  return 0;
-}
+int main(int, char**) { return 0; }
```
Also, this should be a `.compile.pass.cpp`.

philnik777 commented on libcxx/test/std/language.support/support.types/max_align_t.compile.pass.cpp:
```diff
@@ -16,9 +16,13 @@
+static_assert(std::is_trivially_default_constructible<std::max_align_t>::value, "");
 static_assert(std::is_standard_layout<std::max_align_t>::value, "");
-#if TEST_STD_VER <= 17
+#if TEST_STD_VER < 26 // P3247R2
+static_assert(std::is_trivial<std::max_align_t>::value, "");
+#endif
```
Again, no need to test the trait.

philnik777 commented on libcxx/test/std/utilities/meta/meta.trans/meta.trans.other/aligned_storage.pass.cpp:
```diff
@@ -40,10 +40,14 @@ int main(int, char**)
     ASSERT_SAME_TYPE(T1, std::aligned_storage_t<10, 1>);
 #endif
-#if TEST_STD_VER <= 17
+#if TEST_STD_VER < 20 // P0767R1
```
Please avoid unrelated changes.

philnik777 commented on libcxx/include/__type_traits/is_trivial.h:
```diff
@@ -19,12 +19,12 @@
 template <class _Tp>
-struct _LIBCPP_TEMPLATE_VIS _LIBCPP_NO_SPECIALIZATIONS is_trivial : public integral_constant<bool, __is_trivial(_Tp)> {
-};
+struct _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX26 _LIBCPP_NO_SPECIALIZATIONS is_trivial
```
I'd really like to see a message like `Consider using is_trivially_copyable<T>::value && is_trivially_default_constructible<T>::value instead`.

frederick-vs-ja committed:
[libc++] P3247R2: Deprecate `is_trivial(_v)`

Requirements on character-like types are updated unconditionally,
because `basic_string` does requires the default-constructibility.
It might be possible to make `basic_string_view` support classes with
non-public trivial default constructor, but this doesn't seem sensible.

libcxxabi's `ItaniumDemangle.h` is also updated to avoid deprecated
features.

frederick-vs-ja committed:
Merge branch 'main' into p3247r2

frederick-vs-ja committed:
Merge branch 'main' into p3247r2

frederick-vs-ja committed:
Merge branch 'main' into p3247r2
